source/Ship.cpp

```diff
@@ -2645,12 +2739,158 @@ bool Ship::IsParalyzed() const
 
 bool Ship::IsDisabled() const
 {
-	if(!isDisabled)
+	return isDisabled;
+}
+
+
+
+bool Ship::CalculateIsDisabled() const
+{
+	return hull < minimumHull || (!crew && RequiredCrew()) || IsOutOfEnergy();
+}
+
+
+
+// Determine if the ship has any usable weapons.
+bool Ship::IsArmed() const
+{
+	return IsArmed(false);
+}
+
+
+
+// Determine if the ship has any usable weapons against enemies.
+bool Ship::IsArmed(bool includingAntiMissile) const
+{
+	return (includingAntiMissile) ? (maxWeaponRange || hasAntiMissile) : maxWeaponRange;
+}
+
+
+
+// Check for enemies in the in the current system if ship is owned by player.
+bool Ship::IsEnemyInEscortSystem() const
+{
+	if(!IsYours())
 		return false;
+	return isEnemyInEscortSystem;
+}
+
+
 
-	double minimumHull = MinimumHull();
-	bool needsCrew = RequiredCrew() != 0;
-	return (hull < minimumHull || (!crew && needsCrew));
+// If a ship is nearly out of battery energy and is slow to charge or no charge,
+// then consider it low on energy.
+bool Ship::IsEnergyLow() const
+{
+	double frames = 60.;
+	double idleEnergy = frames * GetIdleEnergyPerFrame();
+	double maxEnergy = Attributes().Get("energy capacity");
+	double totalConsumption = frames * GetEnergyConsumptionPerFrame();
+	double currentEnergy = (CanBeCarried() && GetSystem() && !landingPlanet) ? GetCurrentEnergy() : maxEnergy * .75;
+	//double currentEnergy = GetCurrentEnergy();
+	if(totalConsumption < 0.)
+	{
+		double secondsToEmpty = currentEnergy / (-totalConsumption);
+		// how quickly can it charge under no energy load
+		double secondsToFullCharge = (idleEnergy > 0.) ? (maxEnergy - currentEnergy) / idleEnergy : minimumOperatingTime + 1.;
+		if((secondsToEmpty < minimumOperatingTime && secondsToFullCharge > minimumOperatingTime) || (secondsToEmpty < minimumOperatingTime && idleEnergy <= 0.))
```

Comment:
```suggestion
		if(secondsToEmpty < minimumOperatingTime && (idleEnergy <= 0. || secondsToFullCharge > minimumOperatingTime))
```
This keeps the same logic but more optimized